### PR TITLE
trezorActions: Add votingXprivSeed.

### PR DESCRIPTION
**Pull request by JoeGruffins**
Add a function that can query trezor for deterministic bytes to be used
    as the seed for a voting xpriv.

depends on #2869

### CODE REVIEW

File: app/actions/TrezorActions.js [hunk 1]
```diff
@@ -49,6 +49,11 @@ const OP_SSGEN_STR = "bb";
 const OP_SSRTX_STR = "bc";
 const STAKE_REVOCATION = "SSRTX";
 const STAKE_GENERATION = "SSGen";
+// VOTING_XPRIV_KEY and VOTING_XPRIV_VALUE must never change in order to
+// have deterministic voting keys.
+const VOTING_XPRIV_KEY = "Create decred voting xpriv seed bytes?";
```
Comment (JoeGruffins) [inline app/actions/TrezorActions.js:54]: We may be able to get around this by using an IV.

File: app/actions/TrezorActions.js [hunk 2]
```diff
@@ -993,6 +998,46 @@ export const getWalletCreationMasterPubKey = () => async (
...
+        encrypt: true,
+        askOnEncrypt: true,
+        askOnDecrypt: true
+      });
+      return res.payload;
+    });
+    const seed = Buffer.from(ckv.value, "hex");
+    if (seed.length != 32) throw "unusable voting HD pivate key seed bytes";
+    dispatch({ type: TRZ_VOTINGXPRIVSEED_SUCCESS });
+    return seed;
```
Comment (JoeGruffins) [inline app/actions/TrezorActions.js:1035]: I think the best thing to do with the seed is to hand it over to dcrwallet in a new grpc request and have it mathed into the xpriv there, using tested logic.

**Review by JoeGruffins — COMMENTED** [review]
The logic behind the `cipherKeyValue` endpoint is [here](https://github.com/satoshilabs/slips/blob/master/slip-0011.md). Very interested in @matheusd opinion, of course.

File: app/actions/TrezorActions.js (hunk 1 above)
Comment (matheusd) [reply app/actions/TrezorActions.js:54]: An IV has nothing to do with the immutability of the key, it's only used as part of the input to the encryption algorithm.

File: app/actions/TrezorActions.js (hunk 2 above)
Comment (matheusd) [reply app/actions/TrezorActions.js:1035]: Correct.

File: app/actions/TrezorActions.js (hunk 1 above)
Comment (matheusd) [reply app/actions/TrezorActions.js:54]: An IV has nothing to do with the immutability of the key, it's only used as part of the input to the encryption algorithm.

File: app/actions/TrezorActions.js [hunk 3]
```diff
@@ -993,6 +998,46 @@ export const getWalletCreationMasterPubKey = () => async (
...
+    });
+    return;
+  }
+
+  const chainParams = selectors.chainParams(getState());
+
+  try {
+    const ckv = await deviceRun(dispatch, getState, async () => {
+      const res = await session.cipherKeyValue({
+        path: accountPath(WALLET_ACCOUNT, chainParams.HDCoinType),
```
Comment (matheusd) [inline app/actions/TrezorActions.js:1023]: Probably best to use `addressPath` instead of simply `accountPath`

**Review by matheusd — COMMENTED** [review]
It's important to note in the PR/commit that this method was chosen specifically because of instructions by the Trezor team in https://github.com/trezor/trezor-firmware/pull/1249#issuecomment-775208665.

I also think it's important to create a go CLI tool that can recreate this xpriv in software only, so we can be sure that in worst case scenarios (e.g. lost trezor without access to a replacement) a user can recreate its voting xpriv.

My [BIP0039 to DCR Seed](https://github.com/matheusd/bip39-to-dcr-seed) can go from a trezor (+ optional passphrase) to raw dcr seed, so the tool to recreate the xpriv could go from the raw seed to xpriv (or whatever). The important bit is that this must be verified to be working to be offered as a last resort to users.

File: app/actions/TrezorActions.js (hunk 3 above)
Comment (JoeGruffins) [reply app/actions/TrezorActions.js:1023]: Fine with me, but why is that?

File: app/actions/TrezorActions.js (hunk 2 above)
Comment (JoeGruffins) [reply app/actions/TrezorActions.js:1035]: Will do.

File: app/actions/TrezorActions.js (hunk 1 above)
Comment (JoeGruffins) [reply app/actions/TrezorActions.js:54]: I see. The key must not change then.

**Comment by JoeGruffins** [comment]
> I also think it's important to create a go CLI tool that can recreate this xpriv in software only, so we can be sure that in worst case scenarios (e.g. lost trezor without access to a replacement) a user can recreate its voting xpriv.

So, after retrieving, should ask the user to save it, similar to the wallet's seed when creating a wallet?

File: app/actions/TrezorActions.js (hunk 3 above)
Comment (matheusd) [reply app/actions/TrezorActions.js]: We can extract private keys of individual addresses using standard CLI tools if needed.

**Comment by matheusd** [comment]
> > I also think it's important to create a go CLI tool that can recreate this xpriv in software only, so we can be sure that in worst case scenarios (e.g. lost trezor without access to a replacement) a user can recreate its voting xpriv.
> 
> So, after retrieving, should ask the user to save it, similar to the wallet's seed when creating a wallet?

I'm not sure I understood the question, but the idea is that you can pass in the raw dcr hex seed and get back the voting xpriv, so it's a specific process  only used in "disaster recovery" scenarios. Something like:

```
$ bip39-to-dcr-seed
[type seed]
af291933...

$ dcr-voting-xpriv
af231933....
Pm29a00...
```

After that, you can manually import the xpriv in dcrwallet.

File: app/actions/TrezorActions.js (hunk 3 above)
Comment (JoeGruffins) [reply app/actions/TrezorActions.js]: By cli tools do you mean [trezor-cli](https://wiki.trezor.io/Using_trezorctl_commands_with_Trezor)? We cannot extract any private keys afaik.

**Comment by JoeGruffins** [comment]
< I'm not sure I understood the question

I suppose I was thinking we couldn't get the seed again, but it's not like the wallet where you only have the one chance. We can get the seed any time. So nevermind there.

File: app/actions/TrezorActions.js (hunk 3 above)
Comment (matheusd) [reply app/actions/TrezorActions.js]: Standard decred tools if you have the seed imported in a dcrwallet instance

Comment (JoeGruffins) [reply app/actions/TrezorActions.js]: oh, that's right, the trezor seed to dcrwallet, I understand now.

**Comment by alexlyp** [comment]
Any status on this?  It's been in draft since June 2021.

**Comment by JoeGruffins** [comment]
We decided to get dcrwallet changes done first, so waiting for review over there. https://github.com/decred/dcrwallet/pull/2120 and then https://github.com/decred/dcrwallet/pull/1946 and there will need to tie up up some loose ends. So... I can take these down and put them back up later if they are an eyesore.

**Comment by alexlyp** [comment]
Gotcha, thanks for the explanation.  I'll chat with @jrick tomorrow about what he thinks of the dcrwallet portions and then we can go from there.  No need to close these since we're waiting elsewhere for movement.  Thanks!

**Comment by alexlyp** [comment]
Same as from my comment on the Trezor purchase ticket PR.

**Pull request closed without merging**